pkg/docker/config/config.go

```diff
@@ -0,0 +1,262 @@
+}
+
+const (
+	defaultPath       = "/run/user"
+	authCfg           = "containers"
+	authCfgFileName   = "auth.json"
+	dockerCfg         = ".docker"
+	dockerCfgFileName = "config.json"
+	dockerLegacyCfg   = ".dockercfg"
+)
+
+var (
+	// ErrNotLoggedIn is returned for users not logged into a registry
+	// that they are trying to logout of
+	ErrNotLoggedIn = errors.Errorf("not logged in")
+)
+
+// SetAuthentication stores the username and password in the auth.json file
+func SetAuthentication(registry, username, password, pathSetByUser string) error {
+	return modifyJSON(pathSetByUser, func(auths *dockerConfigFile) (bool, error) {
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			return false, setAuthToCredHelper(ch, registry, username, password)
+		}
+
+		creds := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+		newCreds := dockerAuthConfig{Auth: creds}
+		auths.AuthConfigs[registry] = newCreds
+		return true, nil
+	})
+}
+
+// GetAuthentication returns the registry credentials stored in
+// either auth.json file or .docker/config.json
+// If an entry is not found empty strings are returned for the username and password
+func GetAuthentication(registry, pathSetByUser string) (string, string, error) {
+	dockerLegacyPath := filepath.Join(homedir.Get(), dockerLegacyCfg)
+	paths := [3]string{getPathToAuth(pathSetByUser), filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName), dockerLegacyPath}
+
+	for _, path := range paths {
+		legacyFormat := path == dockerLegacyPath
+		username, password, err := findAuthentication(registry, path, legacyFormat)
+		if err != nil {
+			return "", "", err
+		}
+		if username != "" && password != "" {
+			return username, password, nil
+		}
+	}
+	return "", "", nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry, pathSetByUser string) error {
+	return modifyJSON(pathSetByUser, func(auths *dockerConfigFile) (bool, error) {
+		// First try cred helpers.
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			return false, deleteAuthFromCredHelper(ch, registry)
+		}
+
+		if _, ok := auths.AuthConfigs[registry]; ok {
+			delete(auths.AuthConfigs, registry)
+		} else if _, ok := auths.AuthConfigs[normalizeRegistry(registry)]; ok {
+			delete(auths.AuthConfigs, normalizeRegistry(registry))
+		} else {
+			return false, ErrNotLoggedIn
+		}
+		return true, nil
+	})
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth(pathSetByUser string) string {
+	if pathSetByUser != "" {
+		return filepath.Join(pathSetByUser, authCfgFileName)
+	}
+	if os.Getenv("XDG_RUNTIME_DIR") != "" {
+		return filepath.Join(os.Getenv("XDG_RUNTIME_DIR"), authCfg, authCfgFileName)
+	}
+	return filepath.Join(defaultPath, strconv.Itoa(os.Getuid()), authCfg, authCfgFileName)
```

Comment:
(Structuring this as a `runtimeDir := os.Getenv(); if runtimeDir == "" { runtimeDir = filepath.Join(…) }; return filepath.Join(runtimeDir, …)` would more closely match the semantics of `XDG_RUNTIME_DIR`. This is fine as well.)

Reply:
fixed.